cilium-cli: enable websockets for k8s exec

asauber opened the pull request:
Users of Cilium have begun to report that the `cilium sysdump` command encounters errors with the following form:

    error with exec request ... : Upgrade request required: ""

Kubernetes v1.31 made Websocket the default streaming protocol for use by kubectl[1]. As described in KEP-4006[2], the prior protocol in use, SPDY, has been deprecated since 2015, and is no longer supported by many proxies, gateways and loadbalancers. Such proxies will result in clients observing errors like the above in cases where SPDY is not supported.

This PR changes the streaming protocol used by the Cilium CLI for commands such as `cilium sysdump`, to use the new Websocket executor supported by client-go.

It makes use of the fallback mechanism provided by the apimachinery package, to fallback to SPDY in the case that common protocol errors are observed.

[1]: https://kubernetes.io/blog/2024/08/20/websockets-transition/
[2]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4006-transition-spdy-to-websockets

```release-note
cilium-cli: enable websockets for k8s exec
```

asauber committed:
cilium-cli: enable websockets for k8s exec

Users of Cilium have begun to report that the `cilium sysdump` command
encounters errors with the following form:

    error with exec request ... : Upgrade request required: ""

Kubernetes v1.31 made Websocket the default streaming protocol for use
by kubectl[1]. As described in KEP-4006[2], the prior protocol in use,
SPDY, has been deprecated since 2015, and is no longer supported by many
proxies, gateways and loadbalancers. Such proxies will result in clients
observing errors like the above in cases where SPDY is not supported.

This PR changes the streaming protocol used by the Cilium CLI for
commands such as `cilium sysdump`, to use the new Websocket executor
supported by client-go.

It makes use of the fallback mechanism provided by the apimachinery
package, to fallback to SPDY in the case that common protocol errors are
observed.

[1]: https://kubernetes.io/blog/2024/08/20/websockets-transition/
[2]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4006-transition-spdy-to-websockets

Signed-off-by: Andrew Sauber <andrew.sauber@isovalent.com>

bimmlerd commented on cilium-cli/k8s/exec.go:
```diff
@@ -23,6 +26,27 @@ type ExecParameters struct {
+
+	execWebsocket, err := remotecommand.NewWebSocketExecutor(config, "GET", url.String())
+	if err != nil {
+		return nil, fmt.Errorf("error while creating Websocket executor: %w", err)
```
Depending on what the constructor does, this isn't necessarily a hard failure I think. I don't know whether the constructor already checks whether the API server supports the websocket proto, but even if it doesn't today it might at some point - hence I think this should just silently work if the creation of the SPDY one worked, and vice versa.

asauber replied:
Right, since we don't control these constructors, it would be better to log and continue here with some verbosity level. Will make that change.

We also have a report that this codepath is problematic, so I'll be adding another commit to address it https://github.com/cilium/cilium/blob/2603133c543322182ae82b48dfb57837d13cd3a5/cilium-cli/k8s/client.go#L843

asauber replied:
Implemented this

asauber committed:
cilium-cli: enable websocket dialer for k8s portforward

Users of Cilium have reported that the `cilium sysdump` command
encounters errors with the following form:

    failed to collect ...: Upgrade request required

As part of KEP-4006, a protocol translator proxy was implemented to
allow websocket-based connections to the kube-apiserver by kubectl to be
translated to SPDY-based connections to the kubelet by the
kube-apiserver, for the sake of features like kubectl port-forward.

In the case of either protocol for the backend-leg connection, we prefer
to use the websocket-based client connection to avoid errors like the
above.

This PR uses a fallback mechanism which was implemented in the
apimachinery package to support implementations in upgrading to
websocket clients, which was made default in Kubernetes 1.31.

[1]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4006-transition-spdy-to-websockets
[2]: https://github.com/kubernetes/kubernetes/pull/119186

Signed-off-by: Andrew Sauber <andrew.sauber@isovalent.com>
